lib/recommendation-server.js

```diff
@@ -0,0 +1,109 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const EXPORTED_SYMBOLS = ['RecommendationServer']; // eslint-disable-line no-unused-vars
+
+function RecommendationServer(opts) {
+  /*
+  Class providing a connection to a recommendation server and a method for
+  querying it. This class should be initialized once and exported as a
+  singleton to modules that will use it.
+
+  Usage:
+
+  let recommendation = new RecommendationServer({
+    // events is an instance of lib/events
+    events: events,
+    xhr: Cc['@mozilla.org/xmlextras/xmlhttprequest;1'],
+    server: 'http://universal-search.dev',
+    timeout: 1000
+  });
+  recommendation.init();
+  */
+  this.events = opts.events;
+  this.server = opts.server;
+  this.timeout = opts.timeout;
+  // note: xhr service, not xhr instance
+  this.xhr = opts.xhr;
+  this.console = opts.console;
+
+  this.search = this.search.bind(this);
+}
+
+RecommendationServer.prototype = {
+  init: function() {
+    this.request = this.xhr.createInstance();
+    this.request.timeout = this.timeout;
+    this.request.isInFlight = false;
+
+    this.events.subscribe('printable-key', this.search);
+  },
+  destroy: function() {
+    this.request.abort();
+    delete this.request;
+
+    this.events.unsubscribe('printable-key', this.search);
+  },
+  _url: function(query) {
+    return `${this.server}/?q=${encodeURIComponent(query)}`;
+  },
+  search: function(query) {
+    this._search(query).then(data => {
+      this.events.publish('recommendation', data);
+    }, reason => {
+      // TODO: send this information to metrics instead of logging
+      console.error(reason);
+    });
```

Comment:
I'm not sure this is the correct place for the logic. Wouldn't you want this to be in some sort of controller, rather than in a library that exposes it?